x/ibctransfer/keeper/msg_server.go

```diff
@@ -0,0 +1,73 @@
+package keeper
+
+import (
+	"context"
+
+	tokenregistrytypes "github.com/Sifchain/sifnode/x/tokenregistry/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+
+	"github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer/types"
+)
+
+var _ types.MsgServer = msgServer{}
+
+type msgServer struct {
+	bankKeeper          bankkeeper.Keeper
+	tokenRegistryKeeper tokenregistrytypes.Keeper
+	sdkMsgServer        types.MsgServer
+}
+
+// NewMsgServerImpl returns an implementation of the bank MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl( /* bankKeeper, tokenRegistryKeeper */ ) types.MsgServer {
+	return &msgServer{}
+}
+
+// Transfer defines a rpc handler method for MsgTransfer.
+func (srv msgServer) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.MsgTransferResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return nil, err
+	}
+
+	// get token registry entry for sent token
+	registryEntry := srv.tokenRegistryKeeper.GetDenom(ctx, msg.Token.Denom)
+	// check if registry entry has an IBC decimal field
+	if registryEntry.IbcDenom != "" && registryEntry.Decimals > registryEntry.IbcDecimals {
+		// calculate the conversion difference and reduce precision
+		po := registryEntry.Decimals - registryEntry.IbcDecimals
+		decAmount := sdk.NewDecFromInt(msg.Token.Amount)
+		convAmountDec := ReducePrecision(decAmount, po)
+
+		convAmount := sdk.NewIntFromBigInt(convAmountDec.RoundInt().BigInt())
```

Comment:
It might be of value to use TruncateInt() in this case , to be extra careful, we never mint more

Reply:
+1